batch/batch/driver/instance.py

```diff
@@ -39,121 +43,92 @@ async def create(app, name, activation_token, worker_cores_mcpu, zone):
-            'CALL activate_instance(%s, %s, %s);',
-            (self.name, ip_address, timestamp))
-
-        self.instance_pool.adjust_for_remove_instance(self)
-        self._state = 'active'
-        self.ip_address = ip_address
-        self.instance_pool.adjust_for_add_instance(self)
-
-        self.scheduler_state_changed.set()
-
-        return rv['token']
-
-    async def deactivate(self, reason, timestamp=None):
-        if self._state in ('inactive', 'deleted'):
-            return
-
-        if not timestamp:
-            timestamp = time_msecs()
-
-        rv = await self.db.execute_and_fetchone(
-            'CALL deactivate_instance(%s, %s, %s);',
-            (self.name, reason, timestamp))
-
-        if rv['rc'] == 1:
-            log.info(f'{self} with in-memory state {self._state} was already deactivated; {rv}')
-            assert rv['cur_state'] in ('inactive', 'deleted')
+    @state.setter
+    def state(self, new_state):
+        self.instance_pool.n_instances_by_state[self._state] -= 1
+        self.instance_pool.n_instances_by_state[new_state] += 1
+        self._state = new_state
+        if new_state in ('active', 'inactive'):
+            self.instance_pool.scheduler_state_changed.set()
 
-        self.instance_pool.adjust_for_remove_instance(self)
-        self._state = 'inactive'
-        self._free_cores_mcpu = self.cores_mcpu
-        self.instance_pool.adjust_for_add_instance(self)
+    @property
+    def free_cores_mcpu(self):
+        return self._free_cores_mcpu
 
-        # there might be jobs to reschedule
-        self.scheduler_state_changed.set()
+    @free_cores_mcpu.setter
+    def free_cores_mcpu(self, new_free_cores_mcpu):
+        self.instance_pool.live_free_cores_mcpu += new_free_cores_mcpu - self._free_cores_mcpu
+        self._free_cores_mcpu = new_free_cores_mcpu
+        self.healthy_instances_by_free_cores.remove(self)
+        if self.is_healthy():
+            self.healthy_instances_by_free_cores.add(self)
 
-    async def mark_deleted(self, reason, timestamp):
-        if self._state == 'deleted':
-            return
-        if self._state != 'inactive':
-            await self.deactivate(reason, timestamp)
+    @property
+    def last_updated(self):
+        return self._last_updated
 
-        rv = await self.db.execute_and_fetchone(
-            'CALL mark_instance_deleted(%s);',
-            (self.name,))
+    @last_updated.setter
+    def last_updated(self, new_last_updated):
+        self._last_updated = new_last_updated
+        self.instance_pool.instances_by_last_updated.remove(self)
+        self.instance_pool.instances_by_last_updated.add(self)
 
-        if rv['rc'] == 1:
-            log.info(f'{self} with in-memory state {self._state} could not be marked deleted; {rv}')
-            assert rv['cur_state'] == 'deleted'
+    @property
+    def failed_request_count(self):
+        return self._failed_request_count
 
-        self.instance_pool.adjust_for_remove_instance(self)
```

Comment:
The reason why there's always an adjust for remove instance and an add instance is because you can't change the sort key of a sorted collection while the object is in that collection